fix: inherit sidecar security context to ensure termination is capable with pod security context.

justbert opened the pull request.

Commit:
fix: inherit sidecar security context to ensure termination is capable with pod security context.

Commit:
feat: check that api server has pods/ephemeralcontainers.

Commit:
chore: refactor ephemeralcontainers check.

SystemFiles commented on cmd/root.go:
```diff
@@ -201,3 +204,18 @@ func initConfig() {
 }
+
+// Checks the server to ensure it has ephemeral containers.
+func hasEphemeralContainersResource(client *clientset.Clientset) bool {
```
Is this something that maybe should be part of a helpers file? Maybe in `pkg/sidecarterminator`?

justbert replied:
Hmmm! Yeah, maybe! I was keeping in in the root section since it's more of a sanity check and only used once instead of in a loop. 

This is a direct example of our refactor conversation! I pulled it out to simplify the cognitive load.

SystemFiles replied:
Hmm I see. Yeah I suppose since it is only used in a single instance it does allow readers to get an understanding immediately about what is involved in checking for that compatibility without having to search the rest of the codebase.

Good thinking here 😃

SystemFiles commented on cmd/root.go:
```diff
@@ -132,6 +131,10 @@ has completed`,
 		}()
 
+		if !hasEphemeralContainersResource(client) {
+			klog.Fatal(fmt.Errorf("api server does not have pods/ephemeralcontainers resource"))
```
Do we want to use fatal here? I'm wondering if there are instances where Ephemeral containers are not available from the API server, we can use an alternative approach? Perhaps the same strategy that we used prior to the implementation of ephemeral containers?

I am likely missing context here. Just my 2 cents

justbert replied:
Nono, it's something I've been thinking about. It's definitely a possibility, however, the previous implementation relied on the sidecars/containers we are targeting to contain the ability to send some type of signal to kill the main process. It's a solution, though one I don't find as elegant due to the increased attack surface of the containers for such a simple requirement. Ending pods shouldn't require dependencies in the Container itself.

The reason I added this check early on was to fail-fast and essentially notify operators that this solution wouldn't work ASAP, since, depending on the cluster and it's usage, there may not be a sidecar to terminate for days or months.

I was thinking that if someone really requires the `exec kill` functionality they can use v1 of the solution. 

Eventually, I would like to polish this up into an operator!

SystemFiles replied:
Understood! Thank you very much for the added context :)

SystemFiles commented:
https://github.com/StatCan/kubernetes-sidecar-terminator/blob/6d1e1e889f3a59a3ab57ba781227a74b1eb16cdc/pkg/sidecarterminator/terminator.go#L25-L34

Just a nitpick here 😸  , can we add some comments to document the purpose for each field in this struct. Most of them are pretty easily understood, but a little more clarity never hurts 😄

SystemFiles commented:
https://github.com/StatCan/kubernetes-sidecar-terminator/blob/6d1e1e889f3a59a3ab57ba781227a74b1eb16cdc/pkg/sidecarterminator/terminator.go#L38-L49


I like this! Very useful for debugging!

justbert commented:
> https://github.com/StatCan/kubernetes-sidecar-terminator/blob/6d1e1e889f3a59a3ab57ba781227a74b1eb16cdc/pkg/sidecarterminator/terminator.go#L38-L49
> 
> I like this! Very useful for debugging!

Agreed!

SystemFiles reviewed (approved):
Other than a few questions and minor things, this looks great! Approved!

SystemFiles commented:
Looks like I don't have write access so my approval doesn't do much here lol

justbert commented:
> Looks like I don't have write access so my approval doesn't do much here lol

Ooops, let me fix that. Some of these are old and don't have the proper RBAC.

Commit:
chore: remove unused variables and simplify handler use.

justbert commented:
I've added comments to the fields of the struct while also removing unused ones.

I've also taken the time to move around some functions to make reading easier by aligning with flow structure.

I've also removed some of the indirection with the `eventhandler` by removing some instantiation code and only referring to it in the function in which it is being passed to the informer.
